proposals/0000-package-versioning.rst

```diff
@@ -0,0 +1,175 @@
+
+    Hackage revisions are far too powerful a tool to solve a very simple
+    problem; mutable bounds information should be stored separately. Personally:
+    I think the cabal file should only contain known version bounds, and
+    speculative version bounds should be kept separately, but there's no reason
+    why someone couldn't put speculative version bounds in a .cabal file going
+    forward. We simply shouldn't _require_ speculative bounds to be present.
+
+    We took a very pragmatic approach to Stackage, which was to acknowledge what
+    package authors were most likely to be able to comply with, and tailored our
+    approach to that. Any version bounds discussion must have the same
+    philosophy if it is to succeed: we need to acknowledge what authors have
+    been willing to do, and what we can reasonably educate people on
+
+    Authors make mistakes with the PVP (I had two package failures in the past
+    week because of it, even though I was blamed for one of them falsely).
+    Automated tooling will make mistakes with bounds. The PVP itself - even if
+    fully followed - does not guarantee 100% success. The only solution that
+    guarantees a 100% success of a build plan is curation (and even that is
+    brittle due to, e.g., different OSes). So we cannot reject a solution
+    because it won't work in some corner case. We need to minimize the
+    probability of failure, accepting that failure will inevitably happen.
+
+
+[1] https://gist.github.com/snoyberg/f6f10cdbea4b9e22d1b83e490ec59a10
+
+
+Motivation
+----------
+
+The problem addressed in this proposal is to reduce the burden on package
+maintainers who are currently required to update the implicit constraints when
+it becomes evident that subsequently published versions of a given dependency
+are indeed compatible with the current package.
+
+Failing to do so can cause the hackage solver to fail, resulting in not being
+able to use particular packages.
+
+Since the problemmatic constraints are essentially outside the control of the
+original developer, some way of moving the management of these out of the hands
+of the package maintainer is proposed.
+
+In Michael Snoyman's words
+
+* Let people continue adding in weird exceptions and cases that they know for
+  certain
+
+* Automate the thing most people are doing, and thereby ensure (1) even people
+  who don't care about the PVP are providing that info, (2) lower bound bumps
+  are not forgotten either, and (3) it's easier to test upper bound relaxes
+  going forward through automated tooling
+
+
+Proposed Change
+---------------
+
+The details of this section will be fleshed out as part of a discussion/dialogue
+with the interested parties.
+
+One possible solution is put forward
+`here <https://github.com/haskell/cabal/issues/3729>`_, but it probably
+requires too large a change to be practicable.
+
+Other solutions involve automatically setting the required implicit constraints
```

Comment:
To reframe this issue some, lets focus on what we can communicate in a `.cabal` file.

Currently we have no way to express:
> "accept any new versions of this dependency as long as this project compiles and passes tests using them."

This is an extremely basic and important thing to be able to say. It would be worth adding special syntax to `.cabal` files to do so. Happily we don't have to, because we can change what "no upper bound" means in `.cabal` files to cover this case.

Right now "no upper bound" means:
> "all futures versions of this library are valid, even if they don't compile".

This is a kind of crazy thing for someone to want to express (unless they're intentionally trying to break people's solvers) and we shouldn't help them do so. Instead we should remove this meaning of a missing upper bound, and replace it with the earlier one ("as long as this project compiles and passes tests etc.")

Obviously upper bounds will still be very important for projects that depend on the values as well as the types returned by their dependencies (and are unable to capture this in tests). But it seems to me like this would be a clear improvement to what `.cabal` files mean, and that since all of our infrastructure depends on `.cabal` files getting them right should be the our top priority.